sno/working_copy/sqlserver.py

```diff
@@ -0,0 +1,553 @@
+                f"Error creating SQL Server working copy at {path} - non-empty schema already exists"
+            )
+
+    @classmethod
+    def check_valid_path(cls, workdir_path, path):
+        url = urlsplit(path)
+
+        if url.scheme != "mssql":
+            raise click.UsageError(
+                "Invalid postgres URI - Expecting URI in form: mssql://[HOST]/DBNAME/SCHEMA"
+            )
+
+        url_path = url.path
+        path_parts = url_path[1:].split("/", 3) if url_path else []
+
+        suggestion_message = ""
+        if len(path_parts) == 1 and workdir_path is not None:
+            suggested_path = f"/{path_parts[0]}/{cls.default_schema(workdir_path)}"
+            suggested_uri = urlunsplit(
+                [url.scheme, url.netloc, suggested_path, url.query, ""]
+            )
+            suggestion_message = f"\nFor example: {suggested_uri}"
+
+        if len(path_parts) != 2:
+            raise click.UsageError(
+                "Invalid mssql URI - SQL Server working copy requires both dbname and schema:\n"
+                "Expecting URI in form: mssql://[HOST]/DBNAME/SCHEMA"
+                + suggestion_message
+            )
+
+    @classmethod
+    def normalise_path(cls, repo, path):
+        return path
+
+    @classmethod
+    def default_schema(cls, workdir_path):
+        # TODO - promote to superclass
+        stem = workdir_path.stem
+        schema = re.sub("[^a-z0-9]+", "_", stem.lower()) + "_sno"
+        if schema[0].isdigit():
+            schema = "_" + schema
+        return schema
+
+    def __str__(self):
+        p = urlsplit(self.uri)
+        if p.password is not None:
+            nl = p.hostname
+            if p.username is not None:
+                nl = f"{p.username}@{nl}"
+            if p.port is not None:
+                nl += f":{p.port}"
+
+            p._replace(netloc=nl)
+        return p.geturl()
+
+    @contextlib.contextmanager
+    def session(self, bulk=0):
+        """
+        Context manager for GeoPackage DB sessions, yields a connection object inside a transaction
+
+        Calling again yields the _same_ connection, the transaction/etc only happen in the outer one.
+        """
+        L = logging.getLogger(f"{self.__class__.__qualname__}.session")
+
+        if hasattr(self, "_session"):
+            # inner - reuse
+            L.debug("session: existing...")
+            yield self._session
+            L.debug("session: existing/done")
+
+        else:
+            L.debug("session: new...")
+
+            try:
+                self._session = self.sessionmaker()
+
+                # TODO - use tidier syntax for opening transactions from sqlalchemy.
+                yield self._session
+                self._session.commit()
+            except Exception:
+                self._session.rollback()
+                raise
+            finally:
+                self._session.close()
+                del self._session
+                L.debug("session: new/done")
```

Comment:
Can we get rid of this altogether yet? Seems like [SAs session/savepoint handling](https://docs.sqlalchemy.org/en/13/orm/session_transaction.html) achieves the same thing in a more cohesive way?

Reply:
Tidied this up a little bit and moved it into the base class (except the GPKG implementation, which could also use some love).

I tried make further changes in a hopeful way, but it just caused some tests to fail.
Won't change in this PR - changing transaction management isn't straight forward since the nuances are invisible from the code and only show up as a few failing tests. If a few test fails showing that, eg, GPKG locking or transaction rollbacks have stopped working as expected, it's not clear what I should do except put the code back how it was - will require lots of experimentation.

Open questions:
- is `self.session.begin()` the same as `self.session.execute("BEGIN TRANSACTION")` (Answer: no, but not sure of details)
- will the context-manager `with self.session.begin()` automatically rollback transactions if an exception is raised?
- what's the difference between `with self.session.begin()` and `with self.session.transaction`.

Reply:
okay.

yeah, I don't understand sqlalchemy's approach to transaction management. There's huge warnings all over `session.begin()`. They seem to be saying "the session handles transactions for you! just don't even worry about it!" which is a recipe for disaster. The warnings stop _just_ short of saying that `session.begin()` will actually be deprecated or removed at some point, but it seems concerning

> will the context-manager with self.session.begin() automatically rollback transactions if an exception is raised?

pretty sure yes